lms/services/digest/service.py

```diff
@@ -0,0 +1,109 @@
+
+
+class DigestService:
+    """A service for getting digests and emailing them to people."""
+
+    def __init__(
+        self,
+        digest_assistant: DigestAssistant,
+        h_api: HAPI,
+        mailchimp_service: MailchimpService,
+        sender: EmailSender,
+    ):
+        self._assistant = digest_assistant
+        self._h_api = h_api
+        self._mailchimp_service = mailchimp_service
+        self._sender = sender
+
+    def send_emails(
+        self,
+        audience: List[str],
+        updated_after: datetime,
+        updated_before: datetime,
+        override_to_email: str = None,
+    ):
+        """
+        Prepare and send activity digests for users over the given period.
+
+        :param audience: List of H userids to send digests to
+        :param updated_after: Search for activity after this date
+        :param updated_before: Search for activity before this date
+        :param override_to_email: Optional override to the destination email
+            address for testing purposes
+        """
+        for digest in self.get_digests(audience, updated_after, updated_before):
+            template_vars = digest.serialize()
+
+            # Check we have something worth sending, and someone to send it to
+            if template_vars["total_annotations"] and (
+                to_email := override_to_email or digest.audience_user.email
+            ):
+                self._mailchimp_service.send_template(
+                    template_name="instructor-email-digest",
+                    sender=self._sender,
+                    recipient=EmailRecipient(
+                        email=to_email, name=digest.audience_user.display_name
+                    ),
+                    template_vars=template_vars,
+                )
+
+    def get_digests(
+        self, audience: List[str], updated_after: datetime, updated_before: datetime
+    ) -> Iterator[Digest]:
+        """
+        Get activity digests for users over a period.
+
+        :param audience: List of H userids to send digests to
+        :param updated_after: Search for activity after this date
+        :param updated_before: Search for activity before this date
+        """
+        annotations = self._h_api.get_annotations(
+            audience, updated_after, updated_before
+        )
+        if not annotations:
+            return
+
+        # Get courses and allocate the annotations to them
+        courses = self._assistant.get_h_courses(
+            authority_provided_ids={
+                annotation["group"]["authority_provided_id"]
+                for annotation in annotations
+            }
+        )
+        self._add_annotations_to_courses(courses, annotations)
+
+        # Make a mapping from instructors to courses
+        instructors_to_courses = defaultdict(list)
+        for course in courses:
+            for h_userid in course.instructors:
+                instructors_to_courses[h_userid].append(course)
```

Comment:
We do this type of thing twice. It seems like the kind of thing we would be able to accomplish better, but I'm not sure how.